core/trino-main/src/main/java/io/trino/operator/project/PageFieldsToInputParametersRewriter.java

```diff
@@ -42,77 +45,113 @@ private PageFieldsToInputParametersRewriter() {}
     public static Result rewritePageFieldsToInputParameters(RowExpression expression)
     {
         Visitor visitor = new Visitor();
-        RowExpression rewrittenProjection = expression.accept(visitor, null);
-        InputChannels inputChannels = new InputChannels(visitor.getInputChannels());
+        RowExpression rewrittenProjection = expression.accept(visitor, true);
+        InputChannels inputChannels = new InputChannels(visitor.getInputChannels(), visitor.getEagerlyLoadedChannels());
         return new Result(rewrittenProjection, inputChannels);
     }
 
     private static class Visitor
-            implements RowExpressionVisitor<RowExpression, Void>
+            implements RowExpressionVisitor<RowExpression, Boolean>
     {
         private final Map<Integer, Integer> fieldToParameter = new HashMap<>();
         private final List<Integer> inputChannels = new ArrayList<>();
+        private final Set<Integer> eagerlyLoadedChannels = new HashSet<>();
         private int nextParameter;
 
         public List<Integer> getInputChannels()
         {
             return ImmutableList.copyOf(inputChannels);
         }
 
+        public List<Integer> getEagerlyLoadedChannels()
+        {
+            return ImmutableList.copyOf(eagerlyLoadedChannels);
+        }
+
         @Override
-        public RowExpression visitInputReference(InputReferenceExpression reference, Void context)
+        public RowExpression visitInputReference(InputReferenceExpression reference, Boolean unconditionallyEvaluated)
         {
-            int parameter = getParameterForField(reference);
+            int parameter = getParameterForField(reference, unconditionallyEvaluated);
             return field(parameter, reference.getType());
         }
 
-        private Integer getParameterForField(InputReferenceExpression reference)
+        private Integer getParameterForField(InputReferenceExpression reference, boolean unconditionallyEvaluated)
         {
             return fieldToParameter.computeIfAbsent(reference.getField(), field -> {
                 inputChannels.add(field);
+                if (unconditionallyEvaluated) {
+                    eagerlyLoadedChannels.add(reference.getField());
+                }
                 return nextParameter++;
             });
         }
 
         @Override
-        public RowExpression visitCall(CallExpression call, Void context)
+        public RowExpression visitCall(CallExpression call, Boolean unconditionallyEvaluated)
         {
+            boolean containsLambdaExpression = call.getArguments().stream().anyMatch(LambdaDefinitionExpression.class::isInstance);
```

Comment:
I don't fully understand why call expression has special handling of lambdas? Shouldn't lambdas be fully covered via `visitLambda`?

Reply:
When the visitor is traversing lambda expression, the leaf nodes always seem to be ConstantExpression or VariableReferenceExpression. So setting any value for `unconditionallyEvaluated` in `visitLambda` seems to be a no-op.
I think lambda expression is supposed to be always part of a function like `transform`, `reduce` etc. The InputReferenceExpression for the fields potentially accessed by the lambda are found in the initial arguments of CallExpression.

Reply:
I will let this one for @martint to look at.
Could you search for a code that creates `CallExpression` for lambdas and see what are the arguments? According to antrl lambdas are primary expressions, so I'm not sure where `CallExpression` comes from

Reply:
> the leaf nodes always seem to be ConstantExpression or VariableReferenceExpression. So setting any value for unconditionallyEvaluated in visitLambda seems to be a no-op.

You would need to link those references to the corresponding arguments of the call to the higher-order function.

For example, in a hypothetical `apply2` function: `apply2(a, b, (x, y) -> x AND y)`, `b` is conditionally loaded, but `a` isn't due to the conditional nature of `x AND y`.

Reply:
@martint in the current code my intent was to fallback to existing behaviour for lambda expression. The motivating scenario for this PR was to improve efficiency for simple filters and projections (e.g. tpch/q01). Can we skip optimising this scenario or would you still consider this a blocker ?

Reply:
I've added a TODO here now